From bfe5adbe992e57e81368586091006498a57f0619 Mon Sep 17 00:00:00 2001
From: Yanqin Wei <Yanqin.Wei@arm.com>
Date: Tue, 14 Feb 2023 16:01:40 +0800
Subject: [PATCH] include: arch: add macro for mpu memory configuration

Add noncache memory attribute and memory-region-mpu compatible
macro for Armv8r64.

Signed-off-by: Yanqin Wei <Yanqin.Wei@arm.com>
Upstream-Status: Backport [main:2122efee1cfbfd171510f3b5f6f31cfc95cc6aa2]
---
 include/zephyr/arch/arm64/cortex_r/arm_mpu.h | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/include/zephyr/arch/arm64/cortex_r/arm_mpu.h b/include/zephyr/arch/arm64/cortex_r/arm_mpu.h
index e3944a770d..12280f49eb 100644
--- a/include/zephyr/arch/arm64/cortex_r/arm_mpu.h
+++ b/include/zephyr/arch/arm64/cortex_r/arm_mpu.h
@@ -1,7 +1,7 @@
 /*
  * Copyright (c) 2018 Linaro Limited.
  * Copyright (c) 2018 Nordic Semiconductor ASA.
- * Copyright (c) 2021 Arm Limited (or its affiliates). All rights reserved.
+ * Copyright (c) 2021-2023 Arm Limited (or its affiliates). All rights reserved.
  *
  * SPDX-License-Identifier: Apache-2.0
  */
@@ -148,6 +148,14 @@
 		.mair_idx = MPU_MAIR_INDEX_SRAM,		      \
 	}
 
+#define REGION_RAM_NOCACHE_ATTR					      \
+	{							      \
+		/* AP, XN, SH */				      \
+		.rbar = NOT_EXEC | P_RW_U_NA_Msk | NON_SHAREABLE_Msk, \
+		/* Cache-ability */				      \
+		.mair_idx = MPU_MAIR_INDEX_SRAM_NOCACHE,	      \
+	}
+
 #define REGION_RAM_TEXT_ATTR				   \
 	{						   \
 		/* AP, XN, SH */			   \
@@ -219,6 +227,14 @@ struct arm_mpu_config {
 		.attr = _attr,			      \
 	}
 
+#define MPU_REGION_ENTRY_FROM_DTS(_name, _base, _size, _attr) \
+	{						      \
+		.name = _name,				      \
+		.base = _base,				      \
+		.limit = _base + _size,			      \
+		.attr = _attr,				      \
+	}
+
 #define K_MEM_PARTITION_P_RW_U_RW ((k_mem_partition_attr_t) \
 	{(P_RW_U_RW_Msk), MPU_MAIR_INDEX_SRAM})
 #define K_MEM_PARTITION_P_RW_U_NA ((k_mem_partition_attr_t) \
-- 
2.17.1

